_posts/2017-06-08-Rust-1.18.md

```diff
@@ -0,0 +1,200 @@
+the entire struct. No padding is needed. This gives us `1 + 1 + 2 = 4 bytes`.
+
+When designing Rust, stuff like this is why we left the details here undefined.
+It allows for exactly this kind of thing: the compiler can optimize `Suboptimal`
+into `Optimal` automatically. And if you check the sizes of `Suboptimal` and
+`Optimal` on Rust 1.18.0, you'll see that they both have a size of four bytes.
+
+We've been planning this for a while; previous versions of Rust included this
+optimization on the nightly channel, but some people wrote unsafe code that
+assumed the exact details of the representation. We rolled it back while we fixed
+all instances of this that we know about, but if you find some code breaks due to
+this, you should fix it! In the meantime, there's also a flag you can use to control
+this. Imagine we have `Suboptimal` in a file named `foo.rs`, along with a `main`
+function that prints out its size:
+
+```rust
+> rustc foo.rs
+> ./foo
+4
+> rustc foo.rs -Z fuel=foo=0
+optimization-fuel-exhausted: Reorder fields of "Suboptimal"
+> ./foo
+6
+```
+
+This flag is based on an idea called "[Optimization
+fuel](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/07/dfopt-popl10.pdf)":
+
+> It works by giving the optimizer a finite supply of optimization fuel. Each
+> time a rewrite function proposes to replace a node, one unit of fuel is
+> consumed. When the optimizer runs out of fuel, further rewrites are
+> suppressed.
+
+Currently, this is the only optimization that uses the "fuel" concept. By
+setting the fuel to zero, `rustc` will not perform the optimization. This can
+keep your code compiling if you run into a problem, but the longer-term fix is
+to use a `#[repr]` attribute to guarantee a particular layout if you rely on it.
+
+We've been planning on moving `rustdoc` to use a CommonMark compliant markdown
+parser for a long time now. However, just switching over can introduce
+regressions where the CommonMark spec differs from our existing parser, Hoedown.
+As part of the transition plan, [a new flag has been added to
+`rustdoc`](https://github.com/rust-lang/rust/pull/40338), `--enable-commonmark`.
+This will use the new parser instead of the old one. Please give it a try! There's
+no scenario we know of where tweaking your markdown gets identical results on both
+parsers.
+
+Finally, compiling `rustc` itself is now [15%-20% faster](https://github.com/rust-lang/rust/pull/41469). Each commit message in this PR
+goes over the details, but the short of it is that there were some inefficient
+things, and now they've been cleaned up.
+
+See the [detailed release notes][notes] for more.
+
+#### Library stabilizations
+
+Seven new APIs were stabilized this release:
+
+- [`Child::try_wait`] is a non-blocking form of `Child::wait`.
+- [`HashMap::retain`] and  [`HashSet::retain`] bring the `retain` API `Vec<T>` has to these two hash data structures.
+- [`PeekMut::pop`] lets you pop stuff off of a `BinaryHeap<T>`
```

Comment:
You could pop stuff off of `BinaryHeap` already - this lets you pop the top element off of the heap after you've already looked at/messed with it without needing to reorder the heap twice.